### ISSUE TITLE
Any implementation in pytorch?

### ISSUE THREAD

**Issue by John1231983** [voxelmorph/voxelmorph#89]
Hi everyone, I am not family of tensorflow. Has anyone reproduced the project in pytorch? Thanks

**Comment by adalca** [comment]
We have one in the works! Can you please check back, feel free to ping for an update if you don't see an update in a couple of weeks. 

Note that a quick hacky implementation is pretty easy: it's a network (e.g. a UNet ) that you can get from elsewhere, and a spatial transform layer which pytorch now has natively.

**Comment by John1231983** [comment]
Great. Actually, for researcher the PyTorch is perfect and for the product is Tensorflow. I just want to ask some question to make it more clear

1. First, you feed a concatenation of fixed and moving images to the Unet, the last feature size of `Bx3xDxHxW`- called `f_unet` feature or deformation field. We will warp the `f_unet` to the moving image to obtained the moved image using the below code, so Are the U feature in the below figure and deformation field same?
```
grid = f_unet #F.affine_grid(theta, x.size())
moved_image = F.grid_sample(moving_image, grid)
```
 ![stn-arch](https://user-images.githubusercontent.com/24875971/63446196-c32f8f80-c407-11e9-89d3-86aa628a8a47.png)

2. To obtain the deformation field, we will train a loss (simplify as MSE loss) between moved_image and fixed_image (just ignore smooth term). Is it right?

https://pytorch.org/tutorials/intermediate/spatial_transformer_tutorial.html

**Comment by John1231983** [comment]
@adalca : Please correct if I was wrong in pytorch. I just ignore the smooth term. However, the loss does not reduce. 

```
f_unet= unet_feature() # size of Bx3xDxHxW
f_unet = torch.transpose(f_unet,1,4) # Because grid only allows size of BxDxHxWx3
moved = F.grid_sample(moving, f_unet)       
return moved , f_unet # return moved image and deformation field
```

Loss function

```
moved_image, df = voxel_morph(moving, fixed)              
regestation_loss = self.L1(moved_image, fixed)
regestation_loss.backward()
optimizer.step()
```

**Comment by adalca** [comment]
Hi @John1231983 ,

Conceptually I think you got the idea. 

I'm not super familiar with pytorch's built-in `grid_sample` or their spatial transformer tutorial. I do remember that there was an issue with the grid_sample domain space (it requires input in the space of -1, 1 or something similar). Unfortunately I think you'll have to wait for the voxelmorph pytorch release before I can speak a bit more intelligently about this.

**Comment by John1231983** [comment]
That is exactaly the bug what I am facing. I solve it by bellow code. The idea is that cumsum all value of feature map, then divide by shape size for each dim. I hope your work will be soon
```
spatial_gds = torch.rand(1,2,3,3,3)
print (spatial_gds)
_, _, D, H, W = spatial_gds.size()
grids = [torch.cumsum(spatial_gds[:, i, ...], dim=i+1)    for i in range(3)]
#print(grids)
grids_norm = torch.stack([(grid/dim)*2-1. for dim, grid in zip([D,H,W], grids)], dim=4)
print(grids_norm)
```

**Comment by adalca** [comment]
excellent, thanks for the update!

**Comment by John1231983** [comment]
@adalca :Sorry for reopen the issue. The above code is wrong. I have a working code to normalize the df to -1, 1 such as (-1,-1,-1) in the top left. The process is that first we build a linear grid from [0 to (D-1) x 0 to (H-1) x 0 to (W-1)]. Then plus the grid with the last feature map of unet. 
```
grid = grid_linear + f_unet
```
The normalized is done by

```
grid_norm_i= 2* grid_i.clone() / dim_i -1.0 #dim_i is D-1, H-1 or W-1
```

So, we can obtain the grid norm that satisfies the condition of grid_sample function. However, I have a question, we have grid_linear in range of max(H,D,W), while f_unet is unknown range. How we can summarize them together? I mean if we use the weight 1 and 1, then the grid result may highly depend on the f_unet range (for example range of f_unet is -10 to 100, while H=50,W=50 and D= 50 f

**Comment by adalca** [comment]
In the tensorflow implementation, we simply clip the transformation (meshgrid + displacement from `unet`) to the size of the iamges. So if the transformation takes a voxel to -2, we make it 0 -- if it takes a transformation to H+3 we clip it to H.

**Comment by John1231983** [comment]
So, you clip value after summation between meshgrid and displacement (generated by UNet)., right? So, if the displacement is too high, then the summation just depends on the value of displacement. Why not clip displacement to range of size image before summation? For example
```
disp= unet()
dist[0]= tf.clip(disp[0], 0, D)
dist[1]= tf.clip(disp[1], 0, H)
dist[2]= tf.clip(disp[2], 0, W)
dist = dist + meshgrid
```
As I understand, you said as
```
disp= unet()
dist = dist + meshgrid
dist[0]= tf.clip(disp[0], 0, D)
dist[1]= tf.clip(disp[1], 0, H)
dist[2]= tf.clip(disp[2], 0, W)

```

**Comment by adalca** [comment]
In your first case, the final `dist` can still go beyond the image sizes, which you don't want. It also doesn't make much sense to clip the displacement -- the limits (clips) will depend on where the displacement is. a voxel in the middle of the image can move quite a bit, a voxel at the edge of the image is very limited. 

The idea is that based on the displacement, at *each location* you compute the new location of where you'll grab the voxel from, and then make sure that this final absolute value is within the image size.

**Comment by John1231983** [comment]
I understand ! For anyone is looking for the issue. You should you the solution
```
disp= unet()
dist = dist + meshgrid
dist[0]= tf.clip(disp[0], 0, D)
dist[1]= tf.clip(disp[1], 0, H)
dist[2]= tf.clip(disp[2], 0, W)
```

**Comment by adalca** [comment]
If you're going to use `tensorflow`, there's a clean implementation in the function `transform` in https://github.com/adalca/neuron/blob/master/neuron/utils.py

**Comment by John1231983** [comment]
Thanks. It worked now. The voxelmorpho is super fast and competitive with ants. Great job.

For smooth term. Do you think we will smooth f_unet (output of unet network) or f_unet_norm ( normalize with f_unet + meshgrid)? As my opinion, it should smooth to the result of (f_unet+mesh_grid after normalize coordinate)

**Comment by adalca** [comment]
good to hear!

there's different ways to regularize the deformation. In the original voxelmorph paper we penalize the spatial gradient of the displacement (the output of the `unet`). 

Doing the exact same thing after adding the meshgrid is not recommended (since you would penalize the meshgrid) but you could certainly come up with other regularizations that make sense there. 

In the MICCAI paper we build a diffeomorphic layer that guarantees that the deformations are *invertible*, so it's a bit different and yet another way to control the deformation field.

**Comment by zoujinger** [comment]
So when will the voxelmorph pytorch release be available?
I was bothering by the STN for a long time.
Did anyone implement the STN in pytorch?
Thanks so much!

**Comment by adalca** [comment]
@John1231983 @zoujinger please see #109 for a discussion about a pytorch voxelmorph version.
